File: algorithms/0003_longest-substring-without-repeating-characters/java/src/Main.java
import java.util.HashMap;
import java.util.Map;

class Solution {
    public int lengthOfLongestSubstring(String s) {
        int max = 0;
        int start = 0;
        Map<Byte, Integer> m = new HashMap<>();
        byte[] bytes = s.getBytes();
        for (int i = 0; i < bytes.length; i++) {
            byte b = bytes[i];
            Integer objI = m.get(b);
            if (objI != null && objI >= start) {
                start = objI + 1;
            }

            int length = i - start + 1;
            if (length > max) {
                max = length;
            }

            m.put(b, i);
        }
        return max;
    }
}

public class Main {

    public static void main(String[] args) {
        // write your code here
        Solution s = new Solution();
        int max = s.lengthOfLongestSubstring("abcabcbb");
        System.out.println(max);
    }
}

File: algorithms/0003_longest-substring-without-repeating-characters/golang/main.go
package main

import "fmt"

func lengthOfLongestSubstring(s string) int {
	m := make(map[byte]int)
	max := 0
	start := 0
	for i, b := range []byte(s) {
		index, ok := m[b]
		if ok && index >= start {
			start = index + 1
		}

		length := i - start + 1
		if length > max {
			max = length
		}

		m[b] = i
	}
	return max
}

func main() {
	max := lengthOfLongestSubstring("pwwkew")
	fmt.Println(max)
}
